Added PHB-esque fonts thanks to XeLaTeX & other tweaks

jjtParadox opened the pull request:
u/Solbera on reddit made some Creative Commons fonts that are extremely similar to fonts found in the PHB. These fonts are implemented thanks to XeLaTex's fontspec package. If this PR were implemented, this template would have to be used with the XeLaTeX or LuaTeX engines exclusively. The readme would also have to be updated to comply with the CC license used with the fonts, as well as an updated PNG output.

One of the fonts (MrEavesSmallCaps, the font for section headings and monster names) is in an incorrect glyph range, causing LaTeX to not render any glyphs from it correctly. For now, use other fonts and font families instead of this font.

For some reason, adding these fonts tended to break the tcolorboxes, causing strange visual bugs. I fixed all of them except for one; the monster name title. I cannot get a serif small caps bold font to work as the title, so I had to stick with a serif bold until MrEaves is fixed.

Other changes include transparent table rows like what's found in the PHB, an example case of drop caps after a major section heading, and justification similar to what's found in the PHB (left-aligned and no hyphenation).

**UPDATE:** Added compatibility for default LaTeX. XeLaTeX is only required if the user wishes to use the custom fonts.

fverdoja commented:
Nice job! Any way of adding some option to switch from standard fonts (which doesn't require XeLaTeX) and the more accurate ones?

jjtParadox commented:
I'm sure it is possible, but I'm not yet familiar enough with LaTeX to figure out how to do it. My guess would be to add a second parameter to DnD.sty that toggles if the \bookinsanity, \scalysans, etc. commands are registered as new fonts or just aliases of default font families.

jjtParadox commented:
Updated to include most recent changes and added compatibility for the default LaTeX engine.

It was simpler than I expected to do, as I just had to use the `iftex` package and the `\ifPDFTeX` command to register `\scalysans`, `\bookinsanity` etc. as aliases of font families if `pdftex` is used, otherwise register custom fonts.

I had to add some `\ifPDFTeX` switches to the dndcomments and dndpapernotes, as the template's default LaTeX sans-serif font doesn't support bold small caps (this is documented in comments next to the relevant lines of code).

evanbergeron commented:
Sorry for not responding to this sooner. I'd really like to merge this, but have one concern: I don't want to wind up with a mixture of serif fonts. Correct me if I'm wrong, but my impression is that with this PR, subsections wind up in Computer Modern.

If we could resolve that issue, I'll totally game! My genuine thanks to both you and u/Solbera, this is great :)

PS: Thoughts on the leading first letter of sections? Should it be repeated in the smaller text form? Should there be an option to toggle it?

BrianCriswell commented:
@jjtParadox, do you have any input on @evanbergeron's concerns?

jjtParadox commented:
(Fixed a merge conflict, but I think I need to put fonts into some of the new stuff. Don't merge this in just yet)

Alright so there are two sets of fonts used in this PR, one for pdfTeX, one for XeLaTeX. Here's the breakdown:

pdfTeX:
* Leading First Letter: **Bookman**
* Section Headers and Monster Names: **Bookman** (Couldn't decide between Bookman or LMR below)
* Body: **Bookman**
* Table/Comment/Paper Box Header: **Open Sans** (this was used before this PR and was not changed)
* Table/Comment/Paper Box Body: **Latin Modern Sans Serif**
* Monster Body: **Latin Modern Sans Serif**

XeLaTeX:
* Leading First Letter: **Solbera Imitation**
* Section Headers and Monster Names: **Latin Modern Roman** (Solbera's Mr. Eaves doesn't work)
* Body: **Bookinsanity**
* Table/Comment/Paper Box Header: **Scaly Sans**
* Table/Comment/Paper Box Body: **Scaly Sans**
* Monster Body: **Scaly Sans**

The mixture of Serif fonts is straight from the PHB, as shown in [this image](https://drive.google.com/file/d/0B2UyuHLUuCjxQjRUSERsaWpKc1U/view). The pdfTeX version is also as close as possible to the font selection used before the PR and does not use Computer Modern.

You will likely need to decide which font to use for section headers and monster names. I've found that only Bookman supports bold small caps, but I believe normal small caps for Latin Modern Roman make for a fitting substitute that is close to PHB style.

Also, the leading first letter is entirely optional and is up to the user. It is not automatic and is manually included (look at example.tex for an example).

BrianCriswell commented:
Wow, there is a lot going on there, and it looks good. I do not fully understand how LaTeX handles fonts. So some of the following ideas might not work, but I was wondering if you would consider a few changes:
1. I think we are going to eventually move the background images to /lib/img so everything will be in one folder. Could you move the fonts to /lib/font?
2. The font selection is getting a bit messy in some of the package environments because of the lack of support for bold small caps. Could you look at a solution similar to what we are going to do for colors? In other words, could you add a set of \newcommand entries that are similar to what you did for \scalysans but have names for what they do. One such set of commands in the lib/paperbox.sty file might look like:
```
\newcommand\paperboxtitlefont{
    \ifPDFTeX\fontfamily{fosj}\else\scalysans\fi\selectfont\scshape\bfseries\color{black}
}
\newcommand\paperboxfont{\scalysans\selectfont}
```
Part of the paperbox definition then looks like:
```
fonttitle=\paperboxtitlefont,
fontupper=\paperboxfont,
```

This gains 2 things: the definition of the environments such as paperbox are cleaner, and a user can hook in their own font selection and alterations into an environment with \renewcommand.
3. There is much more to going on in the package with fonts now. Would you be willing to move the font code that is not part of an existing environment (such as paperbox) into a new lib/dndfont.sty file?

BrianCriswell commented:
Just wanted to check whether you still wanted me to hold off on merging this.

jjtParadox commented:
Sorry, it's been a busy week! I'll get around to updating this soon and I'll let you know when it can be merged.

BrianCriswell commented:
The differences in the code base are getting to the point that it may be time to close this PR. #75 actually makes it simpler to see the spots where the fonts need to be changed. So your work can be largely preserved by basing a new PR after getting the changes through #75.

Incidently, what do you think of this font for as a replacement for Scala Sans for pdflatex?
<http://www.tug.dk/FontCatalogue/gandhisans/>

BrianCriswell commented:
When you get a chance, go ahead and resubmit this against the current code base.
